### PR TITLE
ENH: Implemented custom mutation / crossover functions for DifferentialEvolution

### PR DESCRIPTION
**Pull request by enn-nafnlaus**
#### Reference issue
None

#### What does this implement/fix?
Implements custom user-specified mutation / crossover functions for DifferentialEvolution

#### Additional information
As per a discussion on the mailing list.  This gives the user full control over the mutation and crossover process by letting them provide their own function:

    strategy_func : fn(candidate, population, rng=None)
        If a custom strategy function is provided, it will be called instead
        of the default function in order to mutate and/or recombine solution
        vectors. It should return a new candidate vector.

### CODE REVIEW

File: scipy/optimize/_differentialevolution.py
Comment (j-bowhay) [inline scipy/optimize/_differentialevolution.py:264]: If merged this will be in 1.12
```suggestion
        .. versionadded:: 1.12.0
```

**Comment by andyfaff** [comment]
@enn-nafnlaus, I can review this today. I did suggest making the PR on a feature branch, not from the main branch of your fork. See https://numpy.org/devdocs/dev/development_workflow.html#making-a-new-feature-branch for good workflow to follow

File: scipy/optimize/tests/test__differential_evolution.py
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py:1598]: Please add tests for when the strategy_func does not return an array of the correct size. You need to test:

```
with raises(RuntimeError, match="strategy_func")
```

File: scipy/optimize/tests/test__differential_evolution.py
Comment (andyfaff) [inline scipy/optimize/tests/test__differential_evolution.py:1587]: - candidate is an integer.
- `assert` is not a function.

```suggestion
            assert 0 < candidate < len(population)
```

In general it's not a good idea to use `np.testing.assert_equal` or `np.array_equal` because of floating point arithmetic. Use `np.testing.assert_allclose` instead.

File: scipy/optimize/tests/test__differential_evolution.py
Comment (andyfaff) [reply scipy/optimize/tests/test__differential_evolution.py:1598]: In offline testing can you see what happens when you supply a function with the wrong signature. What does the error look like? Should we handle that `TypeError` ourselves?

**Review by andyfaff — CHANGES_REQUESTED** [review]

File: scipy/optimize/_differentialevolution.py
Comment (enn-nafnlaus) [reply scipy/optimize/_differentialevolution.py:259]: > BTW, I'm not sure if we'll have a separate keyword argument or include it as part of the `strategy` keyword. pinging @mdhaber for a second opinion.

Not a bad idea - I don't see any issues with implementing it that way, since "strategy" is not used when using strategy_func.

**Comment by enn-nafnlaus** [comment]
> @enn-nafnlaus, I can review this today. I did suggest making the PR on a feature branch, not from the main branch of your fork. See https://numpy.org/devdocs/dev/development_workflow.html#making-a-new-feature-branch for good workflow to follow. Feature branches result in a _lot less_ problems with git. They also allow you to work on more than one feature at a time.

Thanks for that!  Is that advice for the future, or is that a "you need to redo this PR request" sort of thing?

File: scipy/optimize/tests/test__differential_evolution.py
Comment (enn-nafnlaus) [reply scipy/optimize/tests/test__differential_evolution.py:1587]: Should I commit this change even though it takes out the test of whether the arrays match either the (A) initial or (B) new candidate, rather than changing them to assert_allclose?

You'll note that the floating point numbers I chose in the test are all powers of two. Powers of two should always match exactly.

**Comment by andyfaff** [comment]
> Thanks for that! Is that advice for the future, or is that a "you need to redo this PR request" sort of thing?

Advice for the future.

If your git ability is good I would redo on a feature branch. This will then easily allow maintainers to push changes to the feature branch as well. The way I would do it is to squash the commits you've made down into a single commit. Then I would checkout the commit before you started on this PR, create a feature branch, then I would cherry pick the single commit onto that feature branch, push the feature branch to your fork, then remake the PR.
However, if you don't feel comfortable with all of that then I'd leave it.

File: scipy/optimize/_differentialevolution.py
Comment (enn-nafnlaus) [reply scipy/optimize/_differentialevolution.py:1634]: Okay, I'm kind of lost in all the changes. Is this something you've changed and made a request for?  I'll gladly accept the change.  But if I've committed like half the changes you made but then you've not made requests for the other half, and they affect things like indentation and duplication, this is going to get very confusing.  :(

File: scipy/optimize/_differentialevolution.py
Comment (enn-nafnlaus) [reply scipy/optimize/_differentialevolution.py:1603]: See below.

File: scipy/optimize/_differentialevolution.py
Comment (andyfaff) [reply scipy/optimize/_differentialevolution.py:1634]: - Remove lines 1634 to 1641, the lines you added.
- undelete the large bit in red above this section (line numbers 1600-1632)
- I added lines 1619 to 1628, move that to the stat of the method.

(If you'd made this on a feature branch it would be easier for me to make changes directly on the PR, because github often allows maintainers to do that.)

File: scipy/optimize/tests/test__differential_evolution.py
Comment (enn-nafnlaus) [reply scipy/optimize/tests/test__differential_evolution.py:1587]: _"assert is not a function"_ - LOL. I've been using it for years like that and never realized that it's just treating the parentheses as grouping parentheses  ;)  Great to know!

**Comment by andyfaff** [comment]
Actually, let me edit the files directly on GH

**Comment by enn-nafnlaus** [comment]
> If your git ability is good

It is not.  I use SVN in my work.

File: scipy/optimize/tests/test__differential_evolution.py
Comment (enn-nafnlaus) [reply scipy/optimize/tests/test__differential_evolution.py:1598]: > Please add tests for when the strategy_func does not return an array of the correct size. You need to test:
> 
> ```
> with raises(RuntimeError, match="strategy_func")
> ```

Will hold off on this until the above (https://github.com/scipy/scipy/pull/19196#discussion_r1318221398) is resolved.

> In offline testing can you see what happens when you supply a function with the wrong signature. What does the error 
> look like? Should we handle that TypeError ourselves?

Will hold off on this until this (https://github.com/scipy/scipy/pull/19196#discussion_r1317930374) is revolved.

Comment (andyfaff) [reply scipy/optimize/tests/test__differential_evolution.py:1598]: I just resolved it manually.

Comment (enn-nafnlaus) [reply scipy/optimize/tests/test__differential_evolution.py:1598]: > I just resolved it manually.

By all means, commit anything you want. And if anything looks to be too big of a change, feel free to just task it back to me.  :)

**Comment by andyfaff** [comment]
Going to close this PR, and make the changes in #19219. Too difficult to make changes on non-feature branch.

**Pull request closed without merging**
